src/Stratis.Bitcoin/P2P/Peer/NetworkPeerClient.cs

```diff
@@ -145,6 +154,148 @@ public NetworkPeerClient(int Id, TcpClient tcpClient, ILoggerFactory loggerFacto
+        /// <exception cref="FormatException">Thrown if the incoming message is too big.</exception>
+        public async Task<byte[]> ReadMessageAsync(ProtocolVersion protocolVersion, CancellationToken cancellation = default(CancellationToken))
+        {
+            this.logger.LogTrace("({0}:{1})", nameof(protocolVersion), protocolVersion);
+
+            // First find and read the magic.
+            await this.ReadMagicAsync(this.network.MagicBytes, cancellation).ConfigureAwait(false);
+
+            // Then read the header, which is formed of command, length, and possibly also a checksum.
+            int checksumSize = protocolVersion >= ProtocolVersion.MEMPOOL_GD_VERSION ? Message.ChecksumSize : 0;
+            int headerSize = Message.CommandSize + Message.LengthSize + checksumSize;
+
+            byte[] messageHeader = new byte[headerSize];
+            await this.ReadBytesAsync(messageHeader, 0, headerSize, cancellation).ConfigureAwait(false);
+
+            // Then extract the length, which is the message payload size.
+            int lengthOffset = Message.CommandSize;
+            uint length = BitConverter.ToUInt32(messageHeader, lengthOffset);
+
+            // 32 MB limit on message size from Bitcoin Core.
+            if (length > 0x02000000)
+                throw new FormatException("Message payload too big (over 0x02000000 bytes)");
+
+            // Read the payload.
+            int magicLength = this.network.MagicBytes.Length;
+            byte[] message = new byte[magicLength + headerSize + length];
+
+            await this.ReadBytesAsync(message, magicLength + headerSize, (int)length, cancellation).ConfigureAwait(false);
+
+            // And copy the magic and the header to form a complete message.
+            Array.Copy(this.network.MagicBytes, 0, message, 0, this.network.MagicBytes.Length);
+            Array.Copy(messageHeader, 0, message, this.network.MagicBytes.Length, headerSize);
+
+            this.logger.LogTrace("(-):*.{0}={1}", nameof(message.Length), message.Length);
+            return message;
+        }
+
+        /// <summary>
+        /// Seeks and reads the magic value from the connection stream.
+        /// </summary>
+        /// <param name="magic">Magic value that starts the message.</param>
+        /// <param name="cancellation">Cancellation token that allows aborting the read operation.</param>
+        /// <exception cref="OperationCanceledException">Thrown if the operation was cancelled or the end of the stream was reached.</exception>
+        /// <remarks>
+        /// Each networkm message starts with the magic value. If the connection stream is in unknown state,
+        /// the next bytes to read might not be the magic. Therefore we read from the stream until we find the magic value.
+        /// </remarks>
+        public async Task ReadMagicAsync(byte[] magic, CancellationToken cancellation)
```

Comment:
This method came from Network right? so we essentially have duplication of code?

Reply:
yes, yes

Reply:
Hmm wont it be better to improve the code in Network or better yet create a class that lives on network that has the functionality `NetowrkStreamReader`?
Later PR is fine.

Reply:
i am not sure this even belongs to network, you might argue that if you created completely different protocol then it would be part of the network ...